### DISCUSSION TITLE
jbang support / integration

### DISCUSSION THREAD

**Discussion by maxandersen** [konsoletyper/teavm#987] [Ideas]
I became aware of teavm having various cli's published to maven central in https://www.reddit.com/r/java/comments/1hea1ju/teavm_0110_with_support_for_webassembly_gc/ and thought i would continue the conversation over here.

jbang can help make those cli's available to actually use - I kicked that off by creating https://github.com/jbanghub/teavm which lets you do things like:

`jbang teavm-cli@jbanghub/teavm` to run the compiler or if you want to install things, like the devserver do `jbang app install teavm-devserver@jbanghub/teavm` and you can now just run `teavm-devserver`. This will work on windows, mac and linux (and any other OS with java and decent shells available).

That is jbang making teavm more accesible outside just maven/gradle.

Other part is how to best enable jbang projects to build/compile wasm projects.

JBang has a few integration points but not sure which makes sense - probably most `jbang export` or just simple documenting how to actually use the cli above to build ...but I'll need some pointers where it makes sense to start :) 

I

**Comment by maxandersen** [comment]
currently stuck on https://github.com/konsoletyper/teavm/issues/988

**Comment by konsoletyper** [comment]
As a user, how do I use jbang/teavm integration to turn my "Hello, world" app into JavaScript and run it in the browser? Should I do the following:

1. Type JBang command that runs `javac` and produce a class file
2. Type another JBang command, which runs TeaVM, where I specify path to this class file
3. Write HTML file which loads this JavaScript
?

This is not even close to the experience of using JBang when I just type single JBang command and see "Hello, world" in the console. Or it there something else?

**Reply by maxandersen** [reply]
what the final result would look like I'm not sure about yet; but thinking out loud then a first step would being able to do something like:

`jbang export wasm <somejbangapp>` which would make the webassembly which then can be run.

ultimately it shouldn't require much if anymore effort than what using maven/gradle requires.

Maybe a `jbang-teavm` dedicated cli makes sense...not sure.

But initially

"teavm-cli -p `jbang info classpath myapp.java`" might be enough?

What im not sure is what kind of "results"/outputs teavm produces which then can be run?

**Reply by konsoletyper** [reply]
Sure, it would look like this. Also, you need to write TeaVM wasm runtime file, which is not generated by CLI. You can extract it from `teavm-core.jar`. Additionally, it would be good to produce html file as well (with the option for user to disable it). Finally, please note that WebAssembly does not work from local file system and should be served via HTTP.

**Comment by konsoletyper** [comment]
Another thing I noticed is that this integration uses something that perhaps it should not: `org.teavm.cli.TeaVMCBuilderRunner`. It's a CLI for incrementally generating C output.

**Reply by maxandersen** [reply]
yeah, I just added the main methods I could find to show whats possible. Not being able to run the samples made it hard for me testing this yet ;)
